MINOR: update KTable JavaDoc

mjsax opened the pull request.

mjsax committed:
MINOR: update KTable, KGroupedTable, and GlobalKTable JavaDocs

mjsax commented:
@dguy @enothereska @guozhangwang

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/GlobalKTable.java:
```diff
@@ -19,23 +19,28 @@
+ * is fully replicated per {@link KafkaStreams} instance.
+ * Every partition of the underlying topic is consumed by each {@link GlobalKTable}, such that the full set of data is
+ * available in every {@link KafkaStreams} instance.
+ * This provides the ability to perform joins with {@link KStream} and {@link KTable}, without having to repartition the
```
We should actually remove the reference to `KTable` as we don't currently support it

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
- * Each record in this stream is an update on the primary-keyed table with the record key as the primary key.
+ * Each record in this changelog stream is an update on the primary-keyed table with the record key as the primary key.
  * <p>
  * A {@link KTable} is either defined from one or multiple Kafka topics that are consumed message by message or
```
remove "or multiple"
Unless there is something i'm missing a `KTable` is created from exactly one topic unless it is the result of an aggregation

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
-     *
-     * Implementors will need to override toString for keys and values that are not of
-     * type String, Integer etc to get meaningful information.
+     * Print the update records of this {@link KTable} to {@code System.out}.
```
Print the updated...

mjsax replied:
Was intended as is -- records are updates, so its an "update record" (or update-record). "updated" makes sense, too, though. Not sure what the better wording is...

dguy replied:
I see what you mean. I just think "updated records" sounds more natural. But i'm not overly concerned

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
-     * Implementors will need to override toString for keys and values that are not of
-     * type String, Integer etc to get meaningful information.
+     * Print the update records of this {@link KTable} to {@code System.out}.
+     * This function will use the generated name of the parent processor node to label the key/value pairs printed out
```
remove out

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
+     * {@link Integer} etc. to get meaningful information.
+     * <p>
+     * Note that {@code print()} is not applied to the internal state store and only called for each new {@link KTable}
+     * update record.
```
updated

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
-     *                 if not specified the default serde defined in the configs will be used
-     * @param valSerde value serde used to send key-value pairs,
-     *                 if not specified the default serde defined in the configs will be used
+     * Print the update records of this {@link KTable} to {@code System.out}.
```
same as above

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
     /**
-     * Write the elements of this stream to a file at the given path using default serializers and deserializers.
-     * @param filePath name of file to write to
+     * Write the update records of this {@link KTable} to a file at the given path.
```
updated

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
-     * Write the elements of this stream to a file at the given path using default serializers and deserializers.
-     * @param filePath name of file to write to
+     * Write the update records of this {@link KTable} to a file at the given path.
+     * This function will use the generated name of the parent processor node to label the key/value pairs printed out
```
written to the file

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
-     * Write the elements of this stream to a file at the given path.
-     *
-     * @param filePath name of file to write to
+     * Write the update records of this {@link KTable} to a file at the given path.
```
as above

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
 
     /**
-     * Write the elements of this stream to a file at the given path.
+     * Write the update records of this {@link KTable} to a file at the given path.
```
As above

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
+                      final Serde<V> valSerde);
 
     /**
+     * Write the update records of this {@link KTable} to a file at the given path.
```
As above

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
-     * store with the given store name. Also a changelog topic named "${applicationId}-${storeName}-changelog"
-     * will be automatically created in Kafka for failure recovery, where "applicationID"
-     * is specified by the user in {@link org.apache.kafka.streams.StreamsConfig}.
+     * Perform an action on each update record of this {@link KTable}.
```
updated

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
+     * Note that this is a terminal operation that returns void.
+     * <p>
+     * Note that {@code foreach()} is not applied to the internal state store and only called for each new
+     * {@link KTable} update record.
```
new or updated record?

guozhangwang replied:
I am OK with update records.

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
-     * @return a new {@link KTable} that contains the exact same records as this {@link KTable}
+     * Convert this changelog stream to a {@link KStream} using the given {@link KeyValueMapper} to select the new key.
+     * <p>
+     * For example, you can computes the new key as the length of the value string.
```
compute

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
-     * @return a {@link KStream} that contains the same records as this {@link KTable};
-     *         the records are no longer treated as updates on a primary-keyed table,
-     *         but rather as normal key-value pairs in a record stream
+     * Materialize this changelog stream to a topic and creates a new {@link KTable} from the topic using default
```
This doesn't create a new `KTable`. It just writes the data to the topic

mjsax replied:
c&p error from `through`...

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
-     *
-     * @param mapper  @param mapper  the instance of {@link KeyValueMapper}
-     * @param <K1> the new key type
+     * Materialize this changelog stream to a topic and creates a new {@link KTable} from the topic.
```
As above

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
-     *
-     * @return a {@link KTable} that contains join-records for each key and values computed by the given {@link ValueJoiner},
-     *         one for each matched record-pair with the same key
+     * Materialize this changelog stream to a topic and creates a new {@link KTable} from the topic using a customizable
```
As above

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
+     * an internal generated name, and "-repartition" is a fixed suffix.
+     * You can retrieve all generated internal topic names via {@link KafkaStreams#toString()}.
+     * <p>
+     * All data of this {@link KTable} will be redistributed through the repartitioning topic by writing all update
```
updated records to it, and re-reading all updated records from it,

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
+     * an internal generated name, and "-repartition" is a fixed suffix.
+     * You can retrieve all generated internal topic names via {@link KafkaStreams#toString()}.
+     * <p>
+     * All data of this {@link KTable} will be redistributed through the repartitioning topic by writing all update
```
as above

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
+     * The result is an ever updating {@link KTable} that represents the <em>current</em> (i.e., processing time) result
+     * of the join.
+     * <p>
+     * The join is computed by (1) updating the internal state if one {@link KTable} and (2) performing a lookup for a
```
if -> of ?
also add a , before and

mjsax replied:
Why a comma?
To my knowledge, for two things that are listed there is no comma: "A and B"
Only for three or more, there are command: "A, B, C, and D"

Correct my if I am wrong.

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
+     * This happens in a symmetric way, i.e., for each update of either {@code this} or the {@code other} input
+     * {@link KTable} the result gets updated.
+     * <p>
+     * For each {@link KTable} record that finds an corresponding record in the other {@link KTable} the provided
```
an -> a

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
+     * <p>
+     * Repartitioning can happen both input {@link KTable}s.
+     * For this case, all data of a {@link KTable} will be redistributed through the repartitioning topic by writing all
+     * update records to and rereading all update records from it, such that the join input {@link KTable} is
```
as above

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -17,378 +17,831 @@
+     * <p>
+     * Repartitioning can happen both input {@link KTable}s.
+     * For this case, all data of a {@link KTable} will be redistributed through the repartitioning topic by writing all
+     * update records to and rereading all update records from it, such that the join input {@link KTable} is
```
as above

dguy reviewed (changes requested):
Some typos/grammar changes

dguy commented:
Fair enough
On Thu, 26 Jan 2017 at 21:40, Matthias J. Sax <notifications@github.com>
wrote:

> *@mjsax* commented on this pull request.
> ------------------------------
>
> In streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java
> <https://github.com/apache/kafka/pull/2438>:
>
> > -     *
> -     * @param selector      select the grouping key and value to be aggregated
> -     * @param keySerde      key serdes for materializing this stream,
> -     *                      if not specified the default serdes defined in the configs will be used
> -     * @param valueSerde    value serdes for materializing this stream,
> -     *                      if not specified the default serdes defined in the configs will be used
> -     * @param <K1>          the key type of the {@link KGroupedTable}
> -     * @param <V1>          the value type of the {@link KGroupedTable}
> -     *
> -     * @return a {@link KGroupedTable} that contains the re-partitioned records of this {@link KTable}
> +     * Join records of this {@link KTable} with another {@link KTable}'s records using non-windowed inner equi join.
> +     * The join is a primary key join with join attribute {@code thisKTable.key == otherKTable.key}.
> +     * The result is an ever updating {@link KTable} that represents the <em>current</em> (i.e., processing time) result
> +     * of the join.
> +     * <p>
> +     * The join is computed by (1) updating the internal state if one {@link KTable} and (2) performing a lookup for a
>
> Why?
> To my knowledge, for two things that are listed there is no comma: "A and
> B"
> Only for three or more, there are command: "A, B, C, and D"
>
> Correct my if I am wrong.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/kafka/pull/2438>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADWLXXVYJ3srb9ok59OtUXsf3l_3pEDks5rWRKwgaJpZM4LuZIE>
> .
>

mjsax commented:
Addressed @dguy comments. (Not the part about "update records" -- but we can discuss if we want to reword).

Also added `KGroupedTable` and `GlobalKTable`.

Some minor refactoring in `KStream` and `KGroupedStream` (always discovering small issues along the way...)

Call for review @eno @guozhangwang

mjsax commented:
Rebased to resolve conflict.

mjsax committed:
rebased

mjsax committed:
added query hint to GlobalKTable

guozhangwang commented:
Merged to trunk and 0.10.2.
